pandas/formats/format.py

```diff
@@ -2042,7 +2042,12 @@ def get_result_as_array(self):
         """
 
         if self.formatter is not None:
-            return np.array([self.formatter(x) for x in self.values])
+            # na_rep takes precedence over the specified formatter
+            if self.na_rep is not None:
+                return np.array([self.formatter(x) if notnull(x)
+                                 else self.na_rep for x in self.values])
```

Comment:
use notnull on the entire array first to create a mask
format the non null values then fill the nans

see the datetime formatters for how to do this